File: enode/src/main/java/org/enodeframework/commanding/impl/DefaultCommandProcessor.java
package org.enodeframework.commanding.impl;

import com.google.common.base.Strings;
import org.enodeframework.commanding.ICommandProcessor;
import org.enodeframework.commanding.IProcessingCommandHandler;
import org.enodeframework.commanding.ProcessingCommand;
import org.enodeframework.commanding.ProcessingCommandMailbox;
import org.enodeframework.common.io.Task;
import org.enodeframework.common.scheduling.IScheduleService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.Executor;
import java.util.stream.Collectors;

/**
 * @author anruence@gmail.com
 */
public class DefaultCommandProcessor implements ICommandProcessor {
    private static final Logger logger = LoggerFactory.getLogger(DefaultCommandProcessor.class);
    private final ConcurrentMap<String, ProcessingCommandMailbox> mailboxDict;
    private final String taskName;
    private final IProcessingCommandHandler processingCommandHandler;
    private final IScheduleService scheduleService;
    private final Executor executor;
    private int aggregateRootMaxInactiveSeconds = 3600 * 24 * 3;
    private int commandMailBoxPersistenceMaxBatchSize = 1000;
    private int scanExpiredAggregateIntervalMilliseconds = 5000;

    public DefaultCommandProcessor(IProcessingCommandHandler processingCommandHandler, IScheduleService scheduleService, Executor executor) {
        this.processingCommandHandler = processingCommandHandler;
        this.scheduleService = scheduleService;
        this.executor = executor;
        this.mailboxDict = new ConcurrentHashMap<>();
        this.taskName = "CleanInactiveProcessingCommandMailBoxes_" + System.nanoTime() + new Random().nextInt(10000);
    }

    @Override
    public void process(ProcessingCommand processingCommand) {
        String aggregateRootId = processingCommand.getMessage().getAggregateRootId();
        if (Strings.isNullOrEmpty(aggregateRootId)) {
            throw new IllegalArgumentException("aggregateRootId of command cannot be null or empty, commandId:" + processingCommand.getMessage().getId());
        }
        ProcessingCommandMailbox mailbox = mailboxDict.computeIfAbsent(aggregateRootId, x -> new ProcessingCommandMailbox(x, processingCommandHandler, commandMailBoxPersistenceMaxBatchSize, executor));
        long mailboxTryUsingCount = 0L;
        while (!mailbox.tryUsing()) {
            Task.sleep(1);
            mailboxTryUsingCount++;
            if (mailboxTryUsingCount % 10000 == 0) {
                logger.warn("Command mailbox try using count: {}, aggregateRootId: {}", mailboxTryUsingCount, mailbox.getAggregateRootId());
            }
        }
        if (mailbox.isRemoved()) {
            mailbox = new ProcessingCommandMailbox(aggregateRootId, processingCommandHandler, commandMailBoxPersistenceMaxBatchSize, executor);
            mailboxDict.putIfAbsent(aggregateRootId, mailbox);
        }
        mailbox.enqueueMessage(processingCommand);
        mailbox.exitUsing();
    }

    @Override
    public void start() {
        scheduleService.startTask(taskName, this::cleanInactiveMailbox,
                scanExpiredAggregateIntervalMilliseconds,
                scanExpiredAggregateIntervalMilliseconds);
    }

    @Override
    public void stop() {
        scheduleService.stopTask(taskName);
    }

    private boolean isMailBoxAllowRemove(ProcessingCommandMailbox mailbox) {
        return mailbox.isInactive(aggregateRootMaxInactiveSeconds) && !mailbox.isRunning() && mailbox.getTotalUnHandledMessageCount() == 0;
    }

    private void cleanInactiveMailbox() {
        List<Map.Entry<String, ProcessingCommandMailbox>> inactiveList = mailboxDict.entrySet().stream()
                .filter(entry -> isMailBoxAllowRemove(entry.getValue()))
                .collect(Collectors.toList());
        inactiveList.forEach(entry -> {
            if (isMailBoxAllowRemove(entry.getValue())) {
                ProcessingCommandMailbox removed = mailboxDict.remove(entry.getKey());
                if (removed != null) {
                    removed.markAsRemoved();
                    logger.info("Removed inactive command mailbox, aggregateRootId: {}", entry.getKey());
                }
            }
        });
    }

    public int getAggregateRootMaxInactiveSeconds() {
        return aggregateRootMaxInactiveSeconds;
    }

    public void setAggregateRootMaxInactiveSeconds(int aggregateRootMaxInactiveSeconds) {
        this.aggregateRootMaxInactiveSeconds = aggregateRootMaxInactiveSeconds;
    }

    public int getCommandMailBoxPersistenceMaxBatchSize() {
        return commandMailBoxPersistenceMaxBatchSize;
    }

    public void setCommandMailBoxPersistenceMaxBatchSize(int commandMailBoxPersistenceMaxBatchSize) {
        this.commandMailBoxPersistenceMaxBatchSize = commandMailBoxPersistenceMaxBatchSize;
    }

    public int getScanExpiredAggregateIntervalMilliseconds() {
        return scanExpiredAggregateIntervalMilliseconds;
    }

    public void setScanExpiredAggregateIntervalMilliseconds(int scanExpiredAggregateIntervalMilliseconds) {
        this.scanExpiredAggregateIntervalMilliseconds = scanExpiredAggregateIntervalMilliseconds;
    }
}

File: enode/src/main/java/org/enodeframework/commanding/impl/DefaultCommandProcessor.kt
package org.enodeframework.commanding.impl

import kotlinx.coroutines.CoroutineDispatcher
import org.enodeframework.commanding.CommandProcessor
import org.enodeframework.commanding.ProcessingCommand
import org.enodeframework.commanding.ProcessingCommandHandler
import org.enodeframework.commanding.ProcessingCommandMailbox
import org.enodeframework.common.io.Task
import org.enodeframework.common.scheduling.ScheduleService
import org.enodeframework.common.utils.Assert
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap

/**
 * @author anruence@gmail.com
 */
class DefaultCommandProcessor(
    private val processingCommandHandler: ProcessingCommandHandler,
    private val scheduleService: ScheduleService,
    private val coroutineDispatcher: CoroutineDispatcher,
    /**
     * 当使用默认的从内存清理聚合根的服务时，该属性用于配置扫描过期的聚合根的时间间隔，默认为5秒；
     */
    private val scanExpiredAggregateIntervalMilliseconds: Int = 5000,
    /**
     * 当使用默认的MemoryCache时，该属性用于配置聚合根的最长允许的不活跃时间，超过这个时间就认为是过期，就可以从内存清除了；然后下次如果再需要用的时候再重新加载进来；默认为3天；
     */
    private val aggregateRootMaxInactiveSeconds: Int = 3600 * 24 * 3,
    /**
     * CommandMailBox中的命令处理时一次最多处理多少个命令，默认为1000个
     */
    private val commandMailBoxPersistenceMaxBatchSize: Int = 1000
) : CommandProcessor {
    constructor(
        processingCommandHandler: ProcessingCommandHandler,
        scheduleService: ScheduleService,
        coroutineDispatcher: CoroutineDispatcher
    ) : this(
        processingCommandHandler, scheduleService, coroutineDispatcher, 5000, 3600 * 24 * 3, 1000
    )

    private val logger = LoggerFactory.getLogger(DefaultCommandProcessor::class.java)
    private val mailboxDict: ConcurrentMap<String, ProcessingCommandMailbox>
    private val taskName: String
    override fun process(processingCommand: ProcessingCommand) {
        val aggregateRootId = processingCommand.message.aggregateRootId
        Assert.nonNullOrEmpty(
            aggregateRootId,
            "aggregateRootId of command, commandId: ${processingCommand.message.id}"
        )
        var mailbox = mailboxDict.computeIfAbsent(aggregateRootId) { x: String ->
            ProcessingCommandMailbox(
                x, processingCommandHandler, coroutineDispatcher, commandMailBoxPersistenceMaxBatchSize
            )
        }
        var mailboxTryUsingCount = 0L
        while (!mailbox.tryUsing()) {
            Task.sleep(1)
            mailboxTryUsingCount++
            if (mailboxTryUsingCount % 10000 == 0L) {
                logger.warn(
                    "Command mailbox try using count: {}, aggregateRootId: {}",
                    mailboxTryUsingCount,
                    mailbox.aggregateRootId
                )
            }
        }
        if (mailbox.isRemoved()) {
            mailbox = mailboxDict.computeIfAbsent(aggregateRootId) { x: String ->
                ProcessingCommandMailbox(
                    x, processingCommandHandler, coroutineDispatcher, commandMailBoxPersistenceMaxBatchSize
                )
            }
        }
        mailbox.enqueueMessage(processingCommand)
        mailbox.exitUsing()
    }

    override fun start() {
        scheduleService.startTask(
            taskName,
            { cleanInactiveMailbox() },
            scanExpiredAggregateIntervalMilliseconds,
            scanExpiredAggregateIntervalMilliseconds
        )
    }

    override fun stop() {
        scheduleService.stopTask(taskName)
    }

    private fun isMailBoxAllowRemove(mailbox: ProcessingCommandMailbox): Boolean {
        return mailbox.isInactive(aggregateRootMaxInactiveSeconds) && !mailbox.isRunning && mailbox.getTotalUnHandledMessageCount() == 0L
    }

    private fun cleanInactiveMailbox() {
        val inactiveList: List<Map.Entry<String, ProcessingCommandMailbox>> =
            mailboxDict.entries.filter { entry -> isMailBoxAllowRemove(entry.value) }
        inactiveList.forEach { entry: Map.Entry<String, ProcessingCommandMailbox> ->
            if (isMailBoxAllowRemove(entry.value)) {
                val removed = mailboxDict.remove(entry.key)
                if (removed != null) {
                    removed.markAsRemoved()
                    logger.info("Removed inactive command mailbox, aggregateRootId: {}", entry.key)
                }
            }
        }
    }


    init {
        mailboxDict = ConcurrentHashMap()
        taskName = "CleanInactiveProcessingCommandMailBoxes_" + System.nanoTime() + Random().nextInt(10000)
    }
}